### PR TITLE
Introduce Submodule Init and Update

### PR DESCRIPTION
**Pull request by jamill**
Introduce several new submodule methods:
- [X] Init
- [X] Update
- [x] ~~Sync~~
- [x] Tests for the SubmoduleUpdate callbacks

I decided not to include Sync in this PR.

### CODE REVIEW

**Comment by jamill** [comment]
@nulltoken - I moved `Init()`, `Update()` into `SubmoduleCollection`, is this closer to what you were thinking?

File: LibGit2Sharp.Tests/SubmoduleFixture.cs [hunk 1]
```diff
@@ -151,5 +151,129 @@ public void CanStageChangeInSubmoduleViaIndexStageWithOtherPaths(string submodul
...
+
+                Assert.True(submodule.RetrieveStatus().HasFlag(SubmoduleStatus.InWorkDir));
+                Assert.Equal((ObjectId)"480095882d281ed676fe5b863569520e54a7d5c0", submodule.HeadCommitId);
+                Assert.Equal((ObjectId)"480095882d281ed676fe5b863569520e54a7d5c0", submodule.IndexCommitId);
+                Assert.Equal((ObjectId)"480095882d281ed676fe5b863569520e54a7d5c0", submodule.WorkDirCommitId);
+            }
+        }
+
+        [Fact]
+        public void CanUpdateUnitializedSubmodule()
```
Comment (nulltoken) [inline LibGit2Sharp.Tests/SubmoduleFixture.cs]: One of the tests above is named `UpdatingUninitializedSubmoduleThrows` so this test name (which advertizes the exact opposite) may seem confusing?

Maybe `CanUpdateAndInitializeASubModule`?

File: LibGit2Sharp.Tests/SubmoduleFixture.cs [hunk 2]
```diff
@@ -151,5 +151,129 @@ public void CanStageChangeInSubmoduleViaIndexStageWithOtherPaths(string submodul
...
+            {
+                var submodule = repo.Submodules[submoduleName];
+
+                Assert.NotNull(submodule);
+                Assert.True((submodule.RetrieveStatus() & SubmoduleStatus.WorkDirUninitialized) == SubmoduleStatus.WorkDirUninitialized);
+
+                repo.Submodules.Init(submodule.Name, false);
+                repo.Submodules.Update(submodule.Name, new SubmoduleUpdateOptions());
+
+                Assert.True(submodule.RetrieveStatus().HasFlag(SubmoduleStatus.InWorkDir));
```
Comment (nulltoken) [inline LibGit2Sharp.Tests/SubmoduleFixture.cs]: Could we also rely on `.HasFlag()` in the other tests as well?

File: LibGit2Sharp/Core/GitSubmoduleOptions.cs [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    {
+        public uint Version;
+
+        public GitCheckoutOpts CheckoutOptions;
+
+        public GitRemoteCallbacks RemoteCallbacks;
+
+        public CheckoutStrategy CloneCheckoutStrategy;
+
+        public IntPtr Signature;
```
Comment (nulltoken) [inline LibGit2Sharp/Core/GitSubmoduleOptions.cs]: Should we expose this as well and allow the user to provide a `Signature`?

Comment (nulltoken) [inline LibGit2Sharp/Core/GitSubmoduleOptions.cs]: Nah... It's going to be ripped of in the short term with https://github.com/libgit2/libgit2/pull/2808

File: LibGit2Sharp.Tests/SubmoduleFixture.cs [hunk 4]
```diff
@@ -151,5 +151,129 @@ public void CanStageChangeInSubmoduleViaIndexStageWithOtherPaths(string submodul
...
+            string submoduleName = "submodule_target_wd";
+
+            using (var repo = new Repository(path))
+            {
+                var submodule = repo.Submodules[submoduleName];
+
+                Assert.NotNull(submodule);
+                Assert.True((submodule.RetrieveStatus() & SubmoduleStatus.WorkDirUninitialized) == SubmoduleStatus.WorkDirUninitialized);
+
+                Assert.Throws<LibGit2SharpException>(() => repo.Submodules.Update(submodule.Name, new SubmoduleUpdateOptions()));
```
Comment (nulltoken) [inline LibGit2Sharp.Tests/SubmoduleFixture.cs]: Should we create a proper exception (would an distinctive error code exists)?

Comment (jamill) [inline LibGit2Sharp.Tests/SubmoduleFixture.cs]: I think a more specific exception would be nice here, but I don't think lg2 is currently returning a specific error code - I will double check.

**Comment by nulltoken** [comment]
>If we are not exposing these types of operations on the individual objects, should methods like `Branch.Checkout` instead also be on the containing collection?

**Note:** I just realized I've never answered this comment. Sorry. :confused: 

Indeed. Actually, as `Checkout` modifies the external state (workdir, index and potentially the HEAD state), I wouldn't disagree with deprecating it as a method exposed by `Branch`.

**Comment by nulltoken** [comment]
>@nulltoken - I moved `Init()`, `Update()` into `SubmoduleCollection`, is this closer to what you were thinking?

:+1: :heart:

**Comment by jamill** [comment]
@nulltoken - I have added tests to ensure that the checkout callbacks are called as part of submodule update in the last commit. I would have included tests that cover the fetch progress, but I don't think this progress is reported by the local transport. We do at least have some coverage with the `OnUpdateTips`.

I think this covers the methods I would like to add with this PR, so I will take off the WIP tag. I kept the new commit separate for now to call out the new test coverage. I can squash it when ready.

**Comment by nulltoken** [comment]
>I would have included tests that cover the fetch progress, but I don't think this progress is reported by the local transport. We do at least have some coverage with the OnUpdateTips.

@jamill Hmmm. I thought this would have been covered by https://github.com/libgit2/libgit2/pull/2763/files (/cc @carlosmn)

File: LibGit2Sharp/SubmoduleCollection.cs [hunk 5]
```diff
@@ -49,6 +49,70 @@ internal SubmoduleCollection(Repository repo)
         }
 
         /// <summary>
+        /// Initialize specified submodule.
+        /// </summary>
+        /// <param name="name">The name of the submodule to update.</param>
+        /// <param name="overwrite">Overwrite existing entries.</param>
+        public virtual void Init(string name, bool overwrite)
```
Comment (nulltoken) [inline LibGit2Sharp/SubmoduleCollection.cs]: Could you please describe what happens when Initializing over a existing one (and passing overwrite = false)? Throws/silently swallows....

**Comment by nulltoken** [comment]
@jamill :+1: That looks pretty clean to me (besides the nitpicks above).

@dahlbyk @Therzok Any comments?

**Comment by jamill** [comment]
I would have to see why there are no progress reports coming from the update.

**Comment by jamill** [comment]
@nulltoken I looked at the progress callbacks a bit more. It appears that the changes you referenced in libgit2/libgit2#2763 are to add the sideband progress reporting to the `local_download_pack` function in lg2. However, this function does not appear to normally exercised as part of the local clone logic. It looks like all objects are copied over in `clone_local_into` directly through `git_futils_cp_r` (for which there is no progress reporting), and when the remote download function is called, there are no more objects that need to be downloaded, and sideband progress is not called.

**Comment by nulltoken** [comment]
Hmmm. Where did https://github.com/libgit2/libgit2sharp/commit/447a92cb6b8bd8cc71cfcb02fa0290eff562b648 disappear?

**Comment by jamill** [comment]
It got lost during the rebase - it is has returned now.

**Comment by nulltoken** [comment]
@jamill Looks pretty neat to me! Anything you want to add before we merge this in?

**Comment by jamill** [comment]
@nulltoken - nope, I think this should be good to go. Thanks!

**Comment by nulltoken** [comment]
:boom:

**Comment by nulltoken** [comment]
@jamill Just one thing. I feel a bit uneasy that `SubmoduleUpdate` drifted away so much without an test failing. Do you think you may try and add some test coverage in this area in a future PR?

**Pull request merged**